client/state/plugins/installed/selectors.ts

```diff
@@ -0,0 +1,382 @@
+		...getSiteIdsThatHavePlugins( state ).map( ( siteId ) => isRequesting( state, siteId ) ),
+	]
+);
+
+/**
+ * The plugins here differ from the plugin objects found on state.plugins.installed.plugins in that each plugin
+ * object has data from all the different sites on it, whereas on state.plugins.installed.plugins they only have
+ * the state for the site which index they are under. The objects here are the same as those returned from
+ * getAllPluginsIndexedByPluginSlug, except they are indexed by siteId.
+ */
+export const getAllPluginsIndexedBySiteId = createSelector(
+	( state ) => {
+		const allPluginsIndexedByPluginSlug = getAllPluginsIndexedByPluginSlug( state );
+
+		return Object.values( allPluginsIndexedByPluginSlug ).reduce(
+			(
+				pluginsIndexedBySiteId: { [ siteId: number ]: { [ pluginSlug: string ]: Plugin } },
+				plugin
+			) => {
+				Object.keys( plugin.sites )
+					.map( ( siteId ) => Number( siteId ) )
+					.forEach( ( siteId ) => {
+						pluginsIndexedBySiteId[ siteId ] = {
+							...pluginsIndexedBySiteId[ siteId ],
+							[ plugin.slug ]: plugin,
+						};
+					} );
+
+				return pluginsIndexedBySiteId;
+			},
+			{}
+		);
+	},
+	( state ) => [ getAllPluginsIndexedByPluginSlug( state ), getSiteIdsThatHavePlugins( state ) ]
+);
+
+export const getFilteredAndSortedPlugins = createSelector(
+	( state: AppState, siteIds: number[], pluginFilter?: PluginFilter ) => {
+		const allPluginsIndexedBySiteId = getAllPluginsIndexedBySiteId( state );
+
+		// Properties on the objects in allPluginsIndexedBySiteId will be modified and the
+		// selector memoization always returns the same object, so use cloneDeep to avoid
+		// altering it for everyone.
+		const allPluginsForSites: { [ pluginSlug: string ]: Plugin } = cloneDeep(
+			siteIds
+				.map( ( siteId: number ) => allPluginsIndexedBySiteId[ siteId ] )
+				.filter( ( plugin ) => !! plugin )
+				.reduce( ( accumulator, current ) => ( { ...accumulator, ...current } ), {} )
+		);
+
+		// Filter the sites object on the plugins so that only data for the requested siteIds is present
+		for ( const pluginSlug of Object.keys( allPluginsForSites ) ) {
+			allPluginsForSites[ pluginSlug ].sites = pick(
+				allPluginsForSites[ pluginSlug ].sites,
+				siteIds
+			);
+		}
+
+		const pluginList =
+			pluginFilter && _filters[ pluginFilter ]
+				? filter( allPluginsForSites, _filters[ pluginFilter ] )
+				: allPluginsForSites;
+
+		return sortBy( pluginList, ( plugin: Plugin ) => plugin.slug.toLowerCase() ) as Plugin[];
+	},
+	( state: AppState ) => [ getAllPluginsIndexedBySiteId( state ) ],
+	( state: AppState, siteIds: number[], pluginFilter?: PluginFilter ) => {
+		return [ siteIds, pluginFilter ].flat().join( '-' );
+	}
+);
+
+export function getPluginsWithUpdates( state: AppState, siteIds: number[] ) {
+	return filter( getFilteredAndSortedPlugins( state, siteIds, undefined ), _filters.updates ).map(
+		( plugin ) => ( {
+			...plugin,
```

Comment:
Yes, I think I explained a similar situation in another comment. The `version` was coming from `plugin.update.new_version`, and the `update` object is actually a site object and not a plugin object. The `update` object was being lifted onto the plugin from its site objects, and it was whatever the last site in the list was because each one overwrote the previous one. 

This led to some strange behavior and I think it is properly speaking a data normalization error so I eventually decided to leave the site specific data only on the site objects.